go/pkg/services/log_collector/server.go

```diff
@@ -15,37 +15,491 @@
+func (lcs *LogCollectorServer) addItemToInProgress(ctx context.Context, runId string, selector map[string]string) error {
+
+	stateFile, err := lcs.getState()
+	if err != nil {
+		return errors.Wrap(err, "Failed to get state file")
+	}
+
+	loggedItem := LogItem{
+		RunId:         runId,
+		LabelSelector: selector,
+	}
+
+	stateFile.InProgress = append(stateFile.InProgress, loggedItem)
+
+	if err := lcs.writeStateToFile(ctx, stateFile); err != nil {
+		return errors.Wrap(err, "Failed to write state file")
+	}
+
+	return nil
+}
+
+// removeItemFromInProgress removes an item from the `in_progress` list in the state file
+func (lcs *LogCollectorServer) removeItemFromInProgress(ctx context.Context, runId string) error {
+
+	// get state file
+	stateFile, err := lcs.getState()
+	if err != nil {
+		return errors.Wrap(err, "Failed to get state file")
+	}
+
+	// find and remove run id item
+	indexToDelete := 0
+	for i, loggedItem := range stateFile.InProgress {
+		if loggedItem.RunId == runId {
+			indexToDelete = i
+		}
+	}
+
+	// remove item from in progress slice
+	stateFile.InProgress[indexToDelete] = stateFile.InProgress[len(stateFile.InProgress)-1]
+	stateFile.InProgress = stateFile.InProgress[:len(stateFile.InProgress)-1]
+
+	// write state file back
+	if err := lcs.writeStateToFile(ctx, stateFile); err != nil {
+		return errors.Wrap(err, "Failed to write state file")
+	}
+
+	return nil
+}
+
+// getState returns the state file
+func (lcs *LogCollectorServer) getState() (*State, error) {
+
+	// get lock
+	lcs.stateFileLock.Lock()
+	defer lcs.stateFileLock.Unlock()
+
+	// read file
+	stateFilePath := lcs.resolveStateFilePath()
+	if err := common.EnsureFileExists(stateFilePath); err != nil {
+		return nil, errors.Wrap(err, "Failed to ensure state file exists")
+	}
+	stateFileBytes, err := os.ReadFile(stateFilePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to read stateFile")
+	}
+
+	state := &State{}
+
+	if len(stateFileBytes) == 0 {
+
+		// if file is empty, return the empty state instance
+		return state, nil
+	}
+
+	// unmarshal
+	if err := json.Unmarshal(stateFileBytes, state); err != nil {
+		return nil, errors.Wrap(err, "Failed to unmarshal state file")
+	}
+
+	return state, nil
+}
+
+// writeStateToFile writes the state to file
+func (lcs *LogCollectorServer) writeStateToFile(ctx context.Context, state *State) error {
+
+	// marshal state file
+	encodedState, err := json.Marshal(state)
+	if err != nil {
+		return errors.Wrap(err, "Failed to encode state file")
+	}
+
+	// get lock, unlock later
+	lcs.stateFileLock.Lock()
+	defer lcs.stateFileLock.Unlock()
+
+	// write to file
+	stateFilePath := lcs.resolveStateFilePath()
+	return lcs.writeToFile(ctx, stateFilePath, encodedState, false)
+}
+
+// writeToFile writes the given bytes to the given file path
+func (lcs *LogCollectorServer) writeToFile(ctx context.Context, filePath string, content []byte, append bool) error {
```

Comment:
Maybe this should be in `common/utils.go`? Doesn't seem related to the server itself